From 74b3f7ad0423606102b21b337ad573395e58e41f Mon Sep 17 00:00:00 2001
From: David Pennington <dpenning@google.com>
Date: Fri, 7 Aug 2026 13:19:52 -0700
Subject: [PATCH] [M150] Protect against synchronous browser deletion for
 TGDDialog

Original change's description:
> Protect against synchronous browser deletion for TGDDialog
>
> Showing a browser-modal dialog can spin a nested run loop on some
> platforms, during which the owning browser (and thus the
> DeletionDialogController held by BrowserWindowFeatures) may be torn
> down.
>
> Resolve this by using weak ptrs and checking after the synchronous show,
>
> Fixed: 534867485
> Change-Id: Ib4930e288d478a88aa9addc8101755d4e67609c4
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8170480
> Commit-Queue: David Pennington <dpenning@chromium.org>
> Reviewed-by: Darryl James <dljames@chromium.org>
> Cr-Commit-Position: refs/heads/main@{#1671274}

(cherry picked from commit d2299d0dd4583ae879045d7d362dfb5eb64e0fd4)

Bug: 541110771,534867485
Change-Id: Ib4930e288d478a88aa9addc8101755d4e67609c4
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8216473
Commit-Queue: David Pennington <dpenning@chromium.org>
Reviewed-by: Darryl James <dljames@chromium.org>
Cr-Commit-Position: refs/branch-heads/7871@{#4269}
Cr-Branched-From: f542126b8c1b3e80104b26bb05ec830bd1206f29-refs/heads/main@{#1639810}
---
 .../tab_group_deletion_dialog_controller.cc   | 50 ++++++++++++-------
 .../tab_group_deletion_dialog_controller.h    |  5 ++
 ...oup_deletion_dialog_controller_unittest.cc | 16 ++++++
 3 files changed, 52 insertions(+), 19 deletions(-)

diff --git a/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller.cc b/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller.cc
index 6acecd48a85e3..b90db25b31ab6 100644
--- a/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller.cc
+++ b/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller.cc
@@ -352,11 +352,10 @@ DeletionDialogController::DeletionDialogController(
     Profile* profile,
     TabStripModel* tab_strip_model)
     : profile_(CHECK_DEREF(profile)),
-      show_dialog_model_fn_(base::BindRepeating(
-          &DeletionDialogController::CreateDialogFromBrowser,
-          base::Unretained(this),
-          browser)),
       tab_strip_model_(CHECK_DEREF(tab_strip_model)) {
+  show_dialog_model_fn_ =
+      base::BindRepeating(&DeletionDialogController::CreateDialogFromBrowser,
+                          weak_ptr_factory_.GetWeakPtr(), browser);
   tab_strip_model_->AddObserver(this);
 }
 
@@ -382,8 +381,17 @@ bool DeletionDialogController::IsShowingDialog() const {
 void DeletionDialogController::CreateDialogFromBrowser(
     BrowserWindowInterface* browser,
     std::unique_ptr<ui::DialogModel> dialog_model) {
-  widget_ = chrome::ShowBrowserModal(browser->GetBrowserForMigrationOnly(),
-                                     std::move(dialog_model));
+  // Showing a modal dialog can spin a nested run loop on macOS, during which
+  // the owning browser and controller may be torn down. Guard the post-show
+  // write with a weak pointer to avoid a use-after-free write if `this` is
+  // destroyed during ShowBrowserModal.
+  auto weak_ptr = weak_ptr_factory_.GetWeakPtr();
+  views::Widget* widget =
+      chrome::ShowBrowserModal(browser->GetBrowserForMigrationOnly(),
+                               std::move(dialog_model));
+  if (weak_ptr) {
+    weak_ptr->widget_ = widget;
+  }
 }
 
 bool DeletionDialogController::MaybeShowDialog(
@@ -462,6 +470,15 @@ void DeletionDialogController::OnDialogCancel() {
   state_.reset();
 }
 
+void DeletionDialogController::OnCloseAction() {
+  state_.reset();
+}
+
+void DeletionDialogController::OnDialogDestroying() {
+  widget_ = nullptr;
+  state_.reset();
+}
+
 std::unique_ptr<ui::DialogModel> DeletionDialogController::BuildDialogModel(
     const DialogMetadata& metadata) {
   DialogText strings = GetDialogText(GetProfile(), metadata);
@@ -476,25 +493,20 @@ std::unique_ptr<ui::DialogModel> DeletionDialogController::BuildDialogModel(
   dialog_builder.SetTitle(strings.title)
       .AddParagraph(ui::DialogModelLabel(strings.body))
       .AddCancelButton(base::BindOnce(&DeletionDialogController::OnDialogCancel,
-                                      base::Unretained(this)),
+                                      weak_ptr_factory_.GetWeakPtr()),
                        cancel_button_params)
       .AddOkButton(base::BindOnce(&DeletionDialogController::OnDialogOk,
-                                  base::Unretained(this)),
+                                  weak_ptr_factory_.GetWeakPtr()),
                    ui::DialogModel::Button::Params()
                        .SetLabel(strings.ok_text)
                        .SetEnabled(true)
                        .SetId(kDeletionDialogOkButtonId))
-      .SetCloseActionCallback(base::BindOnce(
-          [](DeletionDialogController* dialog_controller) {
-            dialog_controller->state_.reset();
-          },
-          base::Unretained(this)))
-      .SetDialogDestroyingCallback(base::BindOnce(
-          [](DeletionDialogController* dialog_controller) {
-            dialog_controller->widget_ = nullptr;
-            dialog_controller->state_.reset();
-          },
-          base::Unretained(this)));
+      .SetCloseActionCallback(
+          base::BindOnce(&DeletionDialogController::OnCloseAction,
+                         weak_ptr_factory_.GetWeakPtr()))
+      .SetDialogDestroyingCallback(
+          base::BindOnce(&DeletionDialogController::OnDialogDestroying,
+                         weak_ptr_factory_.GetWeakPtr()));
   if (IsDialogSkippable(metadata.type)) {
     dialog_builder.AddCheckbox(
         kDeletionDialogDontAskCheckboxId,
diff --git a/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller.h b/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller.h
index 911999278b45f..6c33c76a3205c 100644
--- a/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller.h
+++ b/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller.h
@@ -11,6 +11,7 @@
 #include "base/functional/callback.h"
 #include "base/memory/raw_ptr.h"
 #include "base/memory/raw_ref.h"
+#include "base/memory/weak_ptr.h"
 #include "chrome/browser/ui/tabs/tab_strip_model_observer.h"
 #include "ui/views/widget/widget_observer.h"
 
@@ -145,6 +146,8 @@ class DeletionDialogController : public TabStripModelObserver {
   // Methods that are bound by the DialogModel to call the callbacks.
   void OnDialogOk();
   void OnDialogCancel();
+  void OnCloseAction();
+  void OnDialogDestroying();
 
   Profile* GetProfile();
 
@@ -162,6 +165,8 @@ class DeletionDialogController : public TabStripModelObserver {
 
   raw_ptr<views::Widget> widget_;
   const raw_ref<TabStripModel> tab_strip_model_;
+
+  base::WeakPtrFactory<DeletionDialogController> weak_ptr_factory_{this};
 };
 
 }  // namespace tab_groups
diff --git a/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller_unittest.cc b/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller_unittest.cc
index 6dfc11a81618c..52cf82a290510 100644
--- a/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller_unittest.cc
+++ b/chrome/browser/ui/tabs/tab_group_deletion_dialog_controller_unittest.cc
@@ -69,3 +69,19 @@ TEST_F(DeletionDialogControllerUnitTest, OnWidgetDestroyedDestroysState) {
   // Make sure the DeletionDialogController has destroyed it's state.
   EXPECT_FALSE(controller_->IsShowingDialog());
 }
+
+TEST_F(DeletionDialogControllerUnitTest, DestroyControllerBeforeDialogModel) {
+  controller_->MaybeShowDialog(
+      DeletionDialogController::DialogMetadata(
+          DeletionDialogController::DialogType::DeleteSingle),
+      base::DoNothing());
+  EXPECT_TRUE(controller_->IsShowingDialog());
+  EXPECT_TRUE(dialog_host_);
+
+  // Destroy the controller while the dialog model is still alive.
+  controller_.reset();
+
+  // Destroying the dialog model host should not crash when triggering callbacks
+  // on the destroyed controller.
+  ui::TestDialogModelHost::DestroyWithoutAction(std::move(dialog_host_));
+}
-- 
2.56.0

